File: src/org/cerion/symcalc/expression/function/integer/Factorial.java
package org.cerion.symcalc.expression.function.integer;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.number.IntegerNum;

public class Factorial extends FunctionExpr {

	public Factorial(Expr ...e) {
		super(FunctionType.FACTORIAL, e);
	}
	
	@Override
	protected Expr evaluate() {
		
		//TODO can work on non-integers
	
		if(get(0).isInteger())
		{
			IntegerNum n = (IntegerNum)get(0);
			return n.factorial();
		}
		
		return this;
	}

}

File: src/org/cerion/symcalc/expression/function/integer/Factorial.kt
package org.cerion.symcalc.expression.function.integer

import org.cerion.symcalc.exception.ValidationException
import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.FunctionExpr
import org.cerion.symcalc.expression.number.NumberType
import org.cerion.symcalc.expression.number.IntegerNum

class Factorial(vararg e: Expr) : FunctionExpr(FunctionExpr.FunctionType.FACTORIAL, *e) {

    override fun evaluate(): Expr {
        //TODO can work on non-integer

        var result = IntegerNum.ONE
        var N = get(0).asInteger()
        while (N > IntegerNum.ONE) {
            result*= N
            N--
        }

        return result
    }

    @Throws(ValidationException::class)
    override fun validate() {
        validateNumberType(0, NumberType.INTEGER)
    }
}
